<a href="https://colab.research.google.com/github/nxzK1/transform_Data_Pandas/blob/main/Ejerc_transf_datos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

# Datos de ventas de clientes.
### Identificar al cliente con mas compras durante esta semana y recompensarlo
### crear nuevas estrategias para la empresa basados en el insight obtenido.

In [2]:
datos_ventas = pd.read_json('/content/dados_vendas_clientes.json')
datos_ventas.head()

dados_vendas
0  {'Data de venda': '06/06/2022', 'Cliente': ['@...
1  {'Data de venda': '07/06/2022', 'Cliente': ['I...
2  {'Data de venda': '08/06/2022', 'Cliente': ['I...
3  {'Data de venda': '09/06/2022', 'Cliente': ['J...
4  {'Data de venda': '10/06/2022', 'Cliente': ['M...

In [3]:
datos_ventas_normalizados = pd.json_normalize(datos_ventas['dados_vendas'])
datos_ventas_normalizados.head()

Data de venda                                            Cliente  \
0    06/06/2022  [@ANA _LUCIA 321, DieGO ARMANDIU 210, DieGO AR...   
1    07/06/2022  [Isabely JOanes 738, Isabely JOanes 738, Isabe...   
2    08/06/2022  [Isabely JOanes 738, JOãO Gabriel 671, Julya m...   
3    09/06/2022  [Julya meireles 914, MaRIA Julia 444, MaRIA Ju...   
4    10/06/2022  [MaRIA Julia 444, PEDRO PASCO 812, Paulo castr...   

                                Valor da compra  
0    [R$ 836,5, R$ 573,33, R$ 392,8, R$ 512,34]  
1  [R$ 825,31, R$ 168,07, R$ 339,18, R$ 314,69]  
2  [R$ 682,05, R$ 386,34, R$ 622,65, R$ 630,79]  
3   [R$ 390,3, R$ 759,16, R$ 334,47, R$ 678,78]  
4  [R$ 314,24, R$ 311,15, R$ 899,16, R$ 885,24]

In [4]:
columnas = list(datos_ventas_normalizados.columns)
columnas

['Data de venda', 'Cliente', 'Valor da compra']

In [5]:
datos_ventas_normalizados = datos_ventas_normalizados.explode(columnas[1:])
datos_ventas_normalizados

Data de venda             Cliente Valor da compra
0    06/06/2022     @ANA _LUCIA 321        R$ 836,5
0    06/06/2022  DieGO ARMANDIU 210       R$ 573,33
0    06/06/2022  DieGO ARMANDIU 210        R$ 392,8
0    06/06/2022  DieGO ARMANDIU 210       R$ 512,34
1    07/06/2022  Isabely JOanes 738       R$ 825,31
1    07/06/2022  Isabely JOanes 738       R$ 168,07
1    07/06/2022  Isabely JOanes 738       R$ 339,18
1    07/06/2022  Isabely JOanes 738       R$ 314,69
2    08/06/2022  Isabely JOanes 738       R$ 682,05
2    08/06/2022    JOãO Gabriel 671       R$ 386,34
2    08/06/2022  Julya meireles 914       R$ 622,65
2    08/06/2022  Julya meireles 914       R$ 630,79
3    09/06/2022  Julya meireles 914        R$ 390,3
3    09/06/2022     MaRIA Julia 444       R$ 759,16
3    09/06/2022     MaRIA Julia 444       R$ 334,47
3    09/06/2022     MaRIA Julia 444       R$ 678,78
4    10/06/2022     MaRIA Julia 444       R$ 314,24
4    10/06/2022     PEDRO PASCO 812       R$ 311,15
4    10/06/2022    Paulo castro 481       R$ 899,16
4    10/06/2022   Thiago fritzz 883       R$ 885,24

In [6]:
datos_ventas_normalizados.reset_index(drop=True, inplace=True)
datos_ventas_normalizados

Data de venda             Cliente Valor da compra
0     06/06/2022     @ANA _LUCIA 321        R$ 836,5
1     06/06/2022  DieGO ARMANDIU 210       R$ 573,33
2     06/06/2022  DieGO ARMANDIU 210        R$ 392,8
3     06/06/2022  DieGO ARMANDIU 210       R$ 512,34
4     07/06/2022  Isabely JOanes 738       R$ 825,31
5     07/06/2022  Isabely JOanes 738       R$ 168,07
6     07/06/2022  Isabely JOanes 738       R$ 339,18
7     07/06/2022  Isabely JOanes 738       R$ 314,69
8     08/06/2022  Isabely JOanes 738       R$ 682,05
9     08/06/2022    JOãO Gabriel 671       R$ 386,34
10    08/06/2022  Julya meireles 914       R$ 622,65
11    08/06/2022  Julya meireles 914       R$ 630,79
12    09/06/2022  Julya meireles 914        R$ 390,3
13    09/06/2022     MaRIA Julia 444       R$ 759,16
14    09/06/2022     MaRIA Julia 444       R$ 334,47
15    09/06/2022     MaRIA Julia 444       R$ 678,78
16    10/06/2022     MaRIA Julia 444       R$ 314,24
17    10/06/2022     PEDRO PASCO 812       R$ 311,15
18    10/06/2022    Paulo castro 481       R$ 899,16
19    10/06/2022   Thiago fritzz 883       R$ 885,24

In [7]:
datos_ventas_normalizados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Data de venda    20 non-null     object
 1   Cliente          20 non-null     object
 2   Valor da compra  20 non-null     object
dtypes: object(3)
memory usage: 612.0+ bytes


In [8]:
datos_ventas_normalizados['Valor da compra'] = datos_ventas_normalizados['Valor da compra'].apply(lambda x: x.replace('R$', '').replace(',', '.').strip())
datos_ventas_normalizados

Data de venda             Cliente Valor da compra
0     06/06/2022     @ANA _LUCIA 321           836.5
1     06/06/2022  DieGO ARMANDIU 210          573.33
2     06/06/2022  DieGO ARMANDIU 210           392.8
3     06/06/2022  DieGO ARMANDIU 210          512.34
4     07/06/2022  Isabely JOanes 738          825.31
5     07/06/2022  Isabely JOanes 738          168.07
6     07/06/2022  Isabely JOanes 738          339.18
7     07/06/2022  Isabely JOanes 738          314.69
8     08/06/2022  Isabely JOanes 738          682.05
9     08/06/2022    JOãO Gabriel 671          386.34
10    08/06/2022  Julya meireles 914          622.65
11    08/06/2022  Julya meireles 914          630.79
12    09/06/2022  Julya meireles 914           390.3
13    09/06/2022     MaRIA Julia 444          759.16
14    09/06/2022     MaRIA Julia 444          334.47
15    09/06/2022     MaRIA Julia 444          678.78
16    10/06/2022     MaRIA Julia 444          314.24
17    10/06/2022     PEDRO PASCO 812          311.15
18    10/06/2022    Paulo castro 481          899.16
19    10/06/2022   Thiago fritzz 883          885.24

In [9]:
datos_ventas_normalizados['Valor da compra'] = datos_ventas_normalizados['Valor da compra'].astype(np.float64)
datos_ventas_normalizados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Data de venda    20 non-null     object 
 1   Cliente          20 non-null     object 
 2   Valor da compra  20 non-null     float64
dtypes: float64(1), object(2)
memory usage: 612.0+ bytes


In [10]:
datos_ventas_normalizados['Cliente'] = datos_ventas_normalizados['Cliente'].str.lower()
datos_ventas_normalizados

Data de venda             Cliente  Valor da compra
0     06/06/2022     @ana _lucia 321           836.50
1     06/06/2022  diego armandiu 210           573.33
2     06/06/2022  diego armandiu 210           392.80
3     06/06/2022  diego armandiu 210           512.34
4     07/06/2022  isabely joanes 738           825.31
5     07/06/2022  isabely joanes 738           168.07
6     07/06/2022  isabely joanes 738           339.18
7     07/06/2022  isabely joanes 738           314.69
8     08/06/2022  isabely joanes 738           682.05
9     08/06/2022    joão gabriel 671           386.34
10    08/06/2022  julya meireles 914           622.65
11    08/06/2022  julya meireles 914           630.79
12    09/06/2022  julya meireles 914           390.30
13    09/06/2022     maria julia 444           759.16
14    09/06/2022     maria julia 444           334.47
15    09/06/2022     maria julia 444           678.78
16    10/06/2022     maria julia 444           314.24
17    10/06/2022     pedro pasco 812           311.15
18    10/06/2022    paulo castro 481           899.16
19    10/06/2022   thiago fritzz 883           885.24

In [11]:
datos_ventas_normalizados['Cliente'] = datos_ventas_normalizados['Cliente'].str.replace('[^a-zA-Z\']', ' ', regex=True).str.strip()
datos_ventas_normalizados

Data de venda         Cliente  Valor da compra
0     06/06/2022      ana  lucia           836.50
1     06/06/2022  diego armandiu           573.33
2     06/06/2022  diego armandiu           392.80
3     06/06/2022  diego armandiu           512.34
4     07/06/2022  isabely joanes           825.31
5     07/06/2022  isabely joanes           168.07
6     07/06/2022  isabely joanes           339.18
7     07/06/2022  isabely joanes           314.69
8     08/06/2022  isabely joanes           682.05
9     08/06/2022    jo o gabriel           386.34
10    08/06/2022  julya meireles           622.65
11    08/06/2022  julya meireles           630.79
12    09/06/2022  julya meireles           390.30
13    09/06/2022     maria julia           759.16
14    09/06/2022     maria julia           334.47
15    09/06/2022     maria julia           678.78
16    10/06/2022     maria julia           314.24
17    10/06/2022     pedro pasco           311.15
18    10/06/2022    paulo castro           899.16
19    10/06/2022   thiago fritzz           885.24

In [12]:
datos_ventas_normalizados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Data de venda    20 non-null     object 
 1   Cliente          20 non-null     object 
 2   Valor da compra  20 non-null     float64
dtypes: float64(1), object(2)
memory usage: 612.0+ bytes


In [13]:
datos_ventas_normalizados['Data de venda'] = pd.to_datetime(datos_ventas_normalizados['Data de venda'])
datos_ventas_normalizados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Data de venda    20 non-null     datetime64[ns]
 1   Cliente          20 non-null     object        
 2   Valor da compra  20 non-null     float64       
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 612.0+ bytes


In [44]:
orden_por_fecha = datos_ventas_normalizados.groupby(['Cliente'])[['Cliente', 'Data de venda']]
orden_por_fecha.head()

Cliente Data de venda
0       ana  lucia    2022-06-06
1   diego armandiu    2022-06-06
2   diego armandiu    2022-06-06
3   diego armandiu    2022-06-06
4   isabely joanes    2022-07-06
5   isabely joanes    2022-07-06
6   isabely joanes    2022-07-06
7   isabely joanes    2022-07-06
8   isabely joanes    2022-08-06
9     jo o gabriel    2022-08-06
10  julya meireles    2022-08-06
11  julya meireles    2022-08-06
12  julya meireles    2022-09-06
13     maria julia    2022-09-06
14     maria julia    2022-09-06
15     maria julia    2022-09-06
16     maria julia    2022-10-06
17     pedro pasco    2022-10-06
18    paulo castro    2022-10-06
19   thiago fritzz    2022-10-06

In [58]:
total_de_compras = datos_ventas_normalizados.groupby(['Cliente'])['Valor da compra'].sum()
total_de_compras = pd.DataFrame(total_de_compras)
total_de_compras.head(10)

Valor da compra
Cliente                        
ana  lucia               836.50
diego armandiu          1478.47
isabely joanes          2329.30
jo o gabriel             386.34
julya meireles          1643.74
maria julia             2086.65
paulo castro             899.16
pedro pasco              311.15
thiago fritzz            885.24

In [60]:
total_de_compras.sort_values('Valor da compra', ascending=False, inplace=True)
total_de_compras.head(10)

Valor da compra
Cliente                        
isabely joanes          2329.30
maria julia             2086.65
julya meireles          1643.74
diego armandiu          1478.47
paulo castro             899.16
thiago fritzz            885.24
ana  lucia               836.50
jo o gabriel             386.34
pedro pasco              311.15

****
La clienta Isabely Joanes fue la que mas compró durante el periodo de ofertas con un valor total de $2329.30 Reales en compras.

La clienta aplica para una recompensa.
****

# Administración de condominios.
### Analizar el retraso de pago de los inquilinos.

In [14]:
datos_condominios = pd.read_json('/content/dados_locacao_imoveis.json')
datos_condominios.head()

dados_locacao
0  {'apartamento': 'A101 (blocoAP)', 'datas_combi...
1  {'apartamento': 'A102 (blocoAP)', 'datas_combi...
2  {'apartamento': 'B201 (blocoAP)', 'datas_combi...
3  {'apartamento': 'B202 (blocoAP)', 'datas_combi...
4  {'apartamento': 'C301 (blocoAP)', 'datas_combi...

In [15]:
datos_condominios_normalizado = pd.json_normalize(datos_condominios['dados_locacao'])
datos_condominios_normalizado.head()

apartamento datas_combinadas_pagamento        datas_de_pagamento  \
0  A101 (blocoAP)   [01/06/2022, 01/07/2022]  [05/06/2022, 03/07/2022]   
1  A102 (blocoAP)   [02/06/2022, 02/07/2022]  [02/06/2022, 06/07/2022]   
2  B201 (blocoAP)   [03/06/2022, 03/07/2022]  [07/06/2022, 03/07/2022]   
3  B202 (blocoAP)   [04/06/2022, 04/07/2022]  [07/06/2022, 05/07/2022]   
4  C301 (blocoAP)   [05/06/2022, 05/07/2022]  [10/06/2022, 09/07/2022]   

                      valor_aluguel  
0  [$ 1000,0 reais, $ 2500,0 reais]  
1  [$ 1100,0 reais, $ 2600,0 reais]  
2  [$ 1200,0 reais, $ 2700,0 reais]  
3  [$ 1300,0 reais, $ 2800,0 reais]  
4  [$ 1400,0 reais, $ 2900,0 reais]

In [16]:
columnas2 = list(datos_condominios_normalizado.columns)
columnas2

['apartamento',
 'datas_combinadas_pagamento',
 'datas_de_pagamento',
 'valor_aluguel']

In [17]:
datos_condominios_normalizado

apartamento datas_combinadas_pagamento        datas_de_pagamento  \
0   A101 (blocoAP)   [01/06/2022, 01/07/2022]  [05/06/2022, 03/07/2022]   
1   A102 (blocoAP)   [02/06/2022, 02/07/2022]  [02/06/2022, 06/07/2022]   
2   B201 (blocoAP)   [03/06/2022, 03/07/2022]  [07/06/2022, 03/07/2022]   
3   B202 (blocoAP)   [04/06/2022, 04/07/2022]  [07/06/2022, 05/07/2022]   
4   C301 (blocoAP)   [05/06/2022, 05/07/2022]  [10/06/2022, 09/07/2022]   
5   C302 (blocoAP)   [06/06/2022, 06/07/2022]  [08/06/2022, 12/07/2022]   
6   D401 (blocoAP)   [07/06/2022, 07/07/2022]  [07/06/2022, 09/07/2022]   
7   D402 (blocoAP)   [08/06/2022, 08/07/2022]  [10/06/2022, 14/07/2022]   
8   E501 (blocoAP)   [09/06/2022, 09/07/2022]  [10/06/2022, 09/07/2022]   
9   E502 (blocoAP)   [10/06/2022, 10/07/2022]  [16/06/2022, 12/07/2022]   
10  F601 (blocoAP)   [11/06/2022, 11/07/2022]  [17/06/2022, 13/07/2022]   
11  F602 (blocoAP)   [12/06/2022, 12/07/2022]  [15/06/2022, 12/07/2022]   
12  G701 (blocoAP)   [13/06/2022, 13/07/2022]  [19/06/2022, 20/07/2022]   
13  G702 (blocoAP)   [14/06/2022, 14/07/2022]  [15/06/2022, 17/07/2022]   
14  H801 (blocoAP)   [15/06/2022, 15/07/2022]  [18/06/2022, 16/07/2022]   

                       valor_aluguel  
0   [$ 1000,0 reais, $ 2500,0 reais]  
1   [$ 1100,0 reais, $ 2600,0 reais]  
2   [$ 1200,0 reais, $ 2700,0 reais]  
3   [$ 1300,0 reais, $ 2800,0 reais]  
4   [$ 1400,0 reais, $ 2900,0 reais]  
5   [$ 1500,0 reais, $ 1200,0 reais]  
6   [$ 1600,0 reais, $ 1300,0 reais]  
7   [$ 1700,0 reais, $ 1400,0 reais]  
8   [$ 1800,0 reais, $ 1500,0 reais]  
9   [$ 1900,0 reais, $ 1600,0 reais]  
10  [$ 2000,0 reais, $ 1000,0 reais]  
11  [$ 2100,0 reais, $ 1100,0 reais]  
12  [$ 2200,0 reais, $ 1200,0 reais]  
13  [$ 2300,0 reais, $ 2100,0 reais]  
14  [$ 2400,0 reais, $ 2200,0 reais]

In [18]:
datos_condominios_normalizado = datos_condominios_normalizado.explode(columnas2[1:])
datos_condominios_normalizado

apartamento datas_combinadas_pagamento datas_de_pagamento  \
0   A101 (blocoAP)                 01/06/2022         05/06/2022   
0   A101 (blocoAP)                 01/07/2022         03/07/2022   
1   A102 (blocoAP)                 02/06/2022         02/06/2022   
1   A102 (blocoAP)                 02/07/2022         06/07/2022   
2   B201 (blocoAP)                 03/06/2022         07/06/2022   
2   B201 (blocoAP)                 03/07/2022         03/07/2022   
3   B202 (blocoAP)                 04/06/2022         07/06/2022   
3   B202 (blocoAP)                 04/07/2022         05/07/2022   
4   C301 (blocoAP)                 05/06/2022         10/06/2022   
4   C301 (blocoAP)                 05/07/2022         09/07/2022   
5   C302 (blocoAP)                 06/06/2022         08/06/2022   
5   C302 (blocoAP)                 06/07/2022         12/07/2022   
6   D401 (blocoAP)                 07/06/2022         07/06/2022   
6   D401 (blocoAP)                 07/07/2022         09/07/2022   
7   D402 (blocoAP)                 08/06/2022         10/06/2022   
7   D402 (blocoAP)                 08/07/2022         14/07/2022   
8   E501 (blocoAP)                 09/06/2022         10/06/2022   
8   E501 (blocoAP)                 09/07/2022         09/07/2022   
9   E502 (blocoAP)                 10/06/2022         16/06/2022   
9   E502 (blocoAP)                 10/07/2022         12/07/2022   
10  F601 (blocoAP)                 11/06/2022         17/06/2022   
10  F601 (blocoAP)                 11/07/2022         13/07/2022   
11  F602 (blocoAP)                 12/06/2022         15/06/2022   
11  F602 (blocoAP)                 12/07/2022         12/07/2022   
12  G701 (blocoAP)                 13/06/2022         19/06/2022   
12  G701 (blocoAP)                 13/07/2022         20/07/2022   
13  G702 (blocoAP)                 14/06/2022         15/06/2022   
13  G702 (blocoAP)                 14/07/2022         17/07/2022   
14  H801 (blocoAP)                 15/06/2022         18/06/2022   
14  H801 (blocoAP)                 15/07/2022         16/07/2022   

     valor_aluguel  
0   $ 1000,0 reais  
0   $ 2500,0 reais  
1   $ 1100,0 reais  
1   $ 2600,0 reais  
2   $ 1200,0 reais  
2   $ 2700,0 reais  
3   $ 1300,0 reais  
3   $ 2800,0 reais  
4   $ 1400,0 reais  
4   $ 2900,0 reais  
5   $ 1500,0 reais  
5   $ 1200,0 reais  
6   $ 1600,0 reais  
6   $ 1300,0 reais  
7   $ 1700,0 reais  
7   $ 1400,0 reais  
8   $ 1800,0 reais  
8   $ 1500,0 reais  
9   $ 1900,0 reais  
9   $ 1600,0 reais  
10  $ 2000,0 reais  
10  $ 1000,0 reais  
11  $ 2100,0 reais  
11  $ 1100,0 reais  
12  $ 2200,0 reais  
12  $ 1200,0 reais  
13  $ 2300,0 reais  
13  $ 2100,0 reais  
14  $ 2400,0 reais  
14  $ 2200,0 reais

In [19]:
datos_condominios_normalizado.reset_index(drop=True, inplace=True)
datos_condominios_normalizado

apartamento datas_combinadas_pagamento datas_de_pagamento  \
0   A101 (blocoAP)                 01/06/2022         05/06/2022   
1   A101 (blocoAP)                 01/07/2022         03/07/2022   
2   A102 (blocoAP)                 02/06/2022         02/06/2022   
3   A102 (blocoAP)                 02/07/2022         06/07/2022   
4   B201 (blocoAP)                 03/06/2022         07/06/2022   
5   B201 (blocoAP)                 03/07/2022         03/07/2022   
6   B202 (blocoAP)                 04/06/2022         07/06/2022   
7   B202 (blocoAP)                 04/07/2022         05/07/2022   
8   C301 (blocoAP)                 05/06/2022         10/06/2022   
9   C301 (blocoAP)                 05/07/2022         09/07/2022   
10  C302 (blocoAP)                 06/06/2022         08/06/2022   
11  C302 (blocoAP)                 06/07/2022         12/07/2022   
12  D401 (blocoAP)                 07/06/2022         07/06/2022   
13  D401 (blocoAP)                 07/07/2022         09/07/2022   
14  D402 (blocoAP)                 08/06/2022         10/06/2022   
15  D402 (blocoAP)                 08/07/2022         14/07/2022   
16  E501 (blocoAP)                 09/06/2022         10/06/2022   
17  E501 (blocoAP)                 09/07/2022         09/07/2022   
18  E502 (blocoAP)                 10/06/2022         16/06/2022   
19  E502 (blocoAP)                 10/07/2022         12/07/2022   
20  F601 (blocoAP)                 11/06/2022         17/06/2022   
21  F601 (blocoAP)                 11/07/2022         13/07/2022   
22  F602 (blocoAP)                 12/06/2022         15/06/2022   
23  F602 (blocoAP)                 12/07/2022         12/07/2022   
24  G701 (blocoAP)                 13/06/2022         19/06/2022   
25  G701 (blocoAP)                 13/07/2022         20/07/2022   
26  G702 (blocoAP)                 14/06/2022         15/06/2022   
27  G702 (blocoAP)                 14/07/2022         17/07/2022   
28  H801 (blocoAP)                 15/06/2022         18/06/2022   
29  H801 (blocoAP)                 15/07/2022         16/07/2022   

     valor_aluguel  
0   $ 1000,0 reais  
1   $ 2500,0 reais  
2   $ 1100,0 reais  
3   $ 2600,0 reais  
4   $ 1200,0 reais  
5   $ 2700,0 reais  
6   $ 1300,0 reais  
7   $ 2800,0 reais  
8   $ 1400,0 reais  
9   $ 2900,0 reais  
10  $ 1500,0 reais  
11  $ 1200,0 reais  
12  $ 1600,0 reais  
13  $ 1300,0 reais  
14  $ 1700,0 reais  
15  $ 1400,0 reais  
16  $ 1800,0 reais  
17  $ 1500,0 reais  
18  $ 1900,0 reais  
19  $ 1600,0 reais  
20  $ 2000,0 reais  
21  $ 1000,0 reais  
22  $ 2100,0 reais  
23  $ 1100,0 reais  
24  $ 2200,0 reais  
25  $ 1200,0 reais  
26  $ 2300,0 reais  
27  $ 2100,0 reais  
28  $ 2400,0 reais  
29  $ 2200,0 reais

In [20]:
datos_condominios_normalizado.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 4 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   apartamento                 30 non-null     object
 1   datas_combinadas_pagamento  30 non-null     object
 2   datas_de_pagamento          30 non-null     object
 3   valor_aluguel               30 non-null     object
dtypes: object(4)
memory usage: 1.1+ KB


In [21]:
datos_condominios_normalizado['valor_aluguel'] = datos_condominios_normalizado['valor_aluguel'].apply(lambda x: x.replace('$ ', '').replace('reais', '').replace(',', '.').strip())
datos_condominios_normalizado

apartamento datas_combinadas_pagamento datas_de_pagamento valor_aluguel
0   A101 (blocoAP)                 01/06/2022         05/06/2022        1000.0
1   A101 (blocoAP)                 01/07/2022         03/07/2022        2500.0
2   A102 (blocoAP)                 02/06/2022         02/06/2022        1100.0
3   A102 (blocoAP)                 02/07/2022         06/07/2022        2600.0
4   B201 (blocoAP)                 03/06/2022         07/06/2022        1200.0
5   B201 (blocoAP)                 03/07/2022         03/07/2022        2700.0
6   B202 (blocoAP)                 04/06/2022         07/06/2022        1300.0
7   B202 (blocoAP)                 04/07/2022         05/07/2022        2800.0
8   C301 (blocoAP)                 05/06/2022         10/06/2022        1400.0
9   C301 (blocoAP)                 05/07/2022         09/07/2022        2900.0
10  C302 (blocoAP)                 06/06/2022         08/06/2022        1500.0
11  C302 (blocoAP)                 06/07/2022         12/07/2022        1200.0
12  D401 (blocoAP)                 07/06/2022         07/06/2022        1600.0
13  D401 (blocoAP)                 07/07/2022         09/07/2022        1300.0
14  D402 (blocoAP)                 08/06/2022         10/06/2022        1700.0
15  D402 (blocoAP)                 08/07/2022         14/07/2022        1400.0
16  E501 (blocoAP)                 09/06/2022         10/06/2022        1800.0
17  E501 (blocoAP)                 09/07/2022         09/07/2022        1500.0
18  E502 (blocoAP)                 10/06/2022         16/06/2022        1900.0
19  E502 (blocoAP)                 10/07/2022         12/07/2022        1600.0
20  F601 (blocoAP)                 11/06/2022         17/06/2022        2000.0
21  F601 (blocoAP)                 11/07/2022         13/07/2022        1000.0
22  F602 (blocoAP)                 12/06/2022         15/06/2022        2100.0
23  F602 (blocoAP)                 12/07/2022         12/07/2022        1100.0
24  G701 (blocoAP)                 13/06/2022         19/06/2022        2200.0
25  G701 (blocoAP)                 13/07/2022         20/07/2022        1200.0
26  G702 (blocoAP)                 14/06/2022         15/06/2022        2300.0
27  G702 (blocoAP)                 14/07/2022         17/07/2022        2100.0
28  H801 (blocoAP)                 15/06/2022         18/06/2022        2400.0
29  H801 (blocoAP)                 15/07/2022         16/07/2022        2200.0

In [22]:
datos_condominios_normalizado['valor_aluguel'] = datos_condominios_normalizado['valor_aluguel'].astype(np.float64)
datos_condominios_normalizado.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 4 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   apartamento                 30 non-null     object 
 1   datas_combinadas_pagamento  30 non-null     object 
 2   datas_de_pagamento          30 non-null     object 
 3   valor_aluguel               30 non-null     float64
dtypes: float64(1), object(3)
memory usage: 1.1+ KB


In [23]:
datos_condominios_normalizado['apartamento'] = datos_condominios_normalizado['apartamento'].str.upper()
datos_condominios_normalizado

apartamento datas_combinadas_pagamento datas_de_pagamento  \
0   A101 (BLOCOAP)                 01/06/2022         05/06/2022   
1   A101 (BLOCOAP)                 01/07/2022         03/07/2022   
2   A102 (BLOCOAP)                 02/06/2022         02/06/2022   
3   A102 (BLOCOAP)                 02/07/2022         06/07/2022   
4   B201 (BLOCOAP)                 03/06/2022         07/06/2022   
5   B201 (BLOCOAP)                 03/07/2022         03/07/2022   
6   B202 (BLOCOAP)                 04/06/2022         07/06/2022   
7   B202 (BLOCOAP)                 04/07/2022         05/07/2022   
8   C301 (BLOCOAP)                 05/06/2022         10/06/2022   
9   C301 (BLOCOAP)                 05/07/2022         09/07/2022   
10  C302 (BLOCOAP)                 06/06/2022         08/06/2022   
11  C302 (BLOCOAP)                 06/07/2022         12/07/2022   
12  D401 (BLOCOAP)                 07/06/2022         07/06/2022   
13  D401 (BLOCOAP)                 07/07/2022         09/07/2022   
14  D402 (BLOCOAP)                 08/06/2022         10/06/2022   
15  D402 (BLOCOAP)                 08/07/2022         14/07/2022   
16  E501 (BLOCOAP)                 09/06/2022         10/06/2022   
17  E501 (BLOCOAP)                 09/07/2022         09/07/2022   
18  E502 (BLOCOAP)                 10/06/2022         16/06/2022   
19  E502 (BLOCOAP)                 10/07/2022         12/07/2022   
20  F601 (BLOCOAP)                 11/06/2022         17/06/2022   
21  F601 (BLOCOAP)                 11/07/2022         13/07/2022   
22  F602 (BLOCOAP)                 12/06/2022         15/06/2022   
23  F602 (BLOCOAP)                 12/07/2022         12/07/2022   
24  G701 (BLOCOAP)                 13/06/2022         19/06/2022   
25  G701 (BLOCOAP)                 13/07/2022         20/07/2022   
26  G702 (BLOCOAP)                 14/06/2022         15/06/2022   
27  G702 (BLOCOAP)                 14/07/2022         17/07/2022   
28  H801 (BLOCOAP)                 15/06/2022         18/06/2022   
29  H801 (BLOCOAP)                 15/07/2022         16/07/2022   

    valor_aluguel  
0          1000.0  
1          2500.0  
2          1100.0  
3          2600.0  
4          1200.0  
5          2700.0  
6          1300.0  
7          2800.0  
8          1400.0  
9          2900.0  
10         1500.0  
11         1200.0  
12         1600.0  
13         1300.0  
14         1700.0  
15         1400.0  
16         1800.0  
17         1500.0  
18         1900.0  
19         1600.0  
20         2000.0  
21         1000.0  
22         2100.0  
23         1100.0  
24         2200.0  
25         1200.0  
26         2300.0  
27         2100.0  
28         2400.0  
29         2200.0

In [24]:
datos_condominios_normalizado['apartamento'] = datos_condominios_normalizado['apartamento'].str.replace('\([^)]*\)', '', regex=True).str.strip()
datos_condominios_normalizado

apartamento datas_combinadas_pagamento datas_de_pagamento  valor_aluguel
0         A101                 01/06/2022         05/06/2022         1000.0
1         A101                 01/07/2022         03/07/2022         2500.0
2         A102                 02/06/2022         02/06/2022         1100.0
3         A102                 02/07/2022         06/07/2022         2600.0
4         B201                 03/06/2022         07/06/2022         1200.0
5         B201                 03/07/2022         03/07/2022         2700.0
6         B202                 04/06/2022         07/06/2022         1300.0
7         B202                 04/07/2022         05/07/2022         2800.0
8         C301                 05/06/2022         10/06/2022         1400.0
9         C301                 05/07/2022         09/07/2022         2900.0
10        C302                 06/06/2022         08/06/2022         1500.0
11        C302                 06/07/2022         12/07/2022         1200.0
12        D401                 07/06/2022         07/06/2022         1600.0
13        D401                 07/07/2022         09/07/2022         1300.0
14        D402                 08/06/2022         10/06/2022         1700.0
15        D402                 08/07/2022         14/07/2022         1400.0
16        E501                 09/06/2022         10/06/2022         1800.0
17        E501                 09/07/2022         09/07/2022         1500.0
18        E502                 10/06/2022         16/06/2022         1900.0
19        E502                 10/07/2022         12/07/2022         1600.0
20        F601                 11/06/2022         17/06/2022         2000.0
21        F601                 11/07/2022         13/07/2022         1000.0
22        F602                 12/06/2022         15/06/2022         2100.0
23        F602                 12/07/2022         12/07/2022         1100.0
24        G701                 13/06/2022         19/06/2022         2200.0
25        G701                 13/07/2022         20/07/2022         1200.0
26        G702                 14/06/2022         15/06/2022         2300.0
27        G702                 14/07/2022         17/07/2022         2100.0
28        H801                 15/06/2022         18/06/2022         2400.0
29        H801                 15/07/2022         16/07/2022         2200.0

In [28]:
datos_condominios_normalizado['datas_combinadas_pagamento'] = pd.to_datetime(datos_condominios_normalizado['datas_combinadas_pagamento'], format='%d/%m/%Y')
datos_condominios_normalizado['datas_de_pagamento'] = pd.to_datetime(datos_condominios_normalizado['datas_de_pagamento'], format='%d/%m/%Y')
datos_condominios_normalizado.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 4 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   apartamento                 30 non-null     object        
 1   datas_combinadas_pagamento  30 non-null     datetime64[ns]
 2   datas_de_pagamento          30 non-null     datetime64[ns]
 3   valor_aluguel               30 non-null     float64       
dtypes: datetime64[ns](2), float64(1), object(1)
memory usage: 1.1+ KB


In [62]:
datos_condominios_normalizado['dias_atraso'] = (datos_condominios_normalizado['datas_de_pagamento'] - datos_condominios_normalizado['datas_combinadas_pagamento']).dt.days
datos_condominios_normalizado

apartamento datas_combinadas_pagamento datas_de_pagamento  valor_aluguel  \
0         A101                 2022-06-01         2022-06-05         1000.0   
1         A101                 2022-07-01         2022-07-03         2500.0   
2         A102                 2022-06-02         2022-06-02         1100.0   
3         A102                 2022-07-02         2022-07-06         2600.0   
4         B201                 2022-06-03         2022-06-07         1200.0   
5         B201                 2022-07-03         2022-07-03         2700.0   
6         B202                 2022-06-04         2022-06-07         1300.0   
7         B202                 2022-07-04         2022-07-05         2800.0   
8         C301                 2022-06-05         2022-06-10         1400.0   
9         C301                 2022-07-05         2022-07-09         2900.0   
10        C302                 2022-06-06         2022-06-08         1500.0   
11        C302                 2022-07-06         2022-07-12         1200.0   
12        D401                 2022-06-07         2022-06-07         1600.0   
13        D401                 2022-07-07         2022-07-09         1300.0   
14        D402                 2022-06-08         2022-06-10         1700.0   
15        D402                 2022-07-08         2022-07-14         1400.0   
16        E501                 2022-06-09         2022-06-10         1800.0   
17        E501                 2022-07-09         2022-07-09         1500.0   
18        E502                 2022-06-10         2022-06-16         1900.0   
19        E502                 2022-07-10         2022-07-12         1600.0   
20        F601                 2022-06-11         2022-06-17         2000.0   
21        F601                 2022-07-11         2022-07-13         1000.0   
22        F602                 2022-06-12         2022-06-15         2100.0   
23        F602                 2022-07-12         2022-07-12         1100.0   
24        G701                 2022-06-13         2022-06-19         2200.0   
25        G701                 2022-07-13         2022-07-20         1200.0   
26        G702                 2022-06-14         2022-06-15         2300.0   
27        G702                 2022-07-14         2022-07-17         2100.0   
28        H801                 2022-06-15         2022-06-18         2400.0   
29        H801                 2022-07-15         2022-07-16         2200.0   

    dias_atraso  
0             4  
1             2  
2             0  
3             4  
4             4  
5             0  
6             3  
7             1  
8             5  
9             4  
10            2  
11            6  
12            0  
13            2  
14            2  
15            6  
16            1  
17            0  
18            6  
19            2  
20            6  
21            2  
22            3  
23            0  
24            6  
25            7  
26            1  
27            3  
28            3  
29            1

In [67]:
media_de_atraso = datos_condominios_normalizado.groupby(['apartamento'])['dias_atraso'].mean()
media_de_atraso = pd.DataFrame(media_de_atraso)
media_de_atraso

dias_atraso
apartamento             
A101                 3.0
A102                 2.0
B201                 2.0
B202                 2.0
C301                 4.5
C302                 4.0
D401                 1.0
D402                 4.0
E501                 0.5
E502                 4.0
F601                 4.0
F602                 1.5
G701                 6.5
G702                 2.0
H801                 2.0

In [70]:
media_de_atraso.sort_values('dias_atraso', ascending=False, inplace=True)
media_de_atraso

dias_atraso
apartamento             
G701                 6.5
C301                 4.5
C302                 4.0
E502                 4.0
F601                 4.0
D402                 4.0
A101                 3.0
H801                 2.0
B202                 2.0
A102                 2.0
B201                 2.0
G702                 2.0
F602                 1.5
D401                 1.0
E501                 0.5

****
El apartamento G701 promedia un atraso del pago de hasta 6 días.
Siendo este el inquilino que más demora en pagar el alquiler.
****

In [71]:
media_de_atraso.mean()

dias_atraso    2.866667
dtype: float64

****
Se promedia un atraso de 3 días aproximadamente en el pago del alquiler por inquilino deudor.
****